Update to vertx 3.8.1

cescoffier opened the pull request:
This PR updates the Vert.x version to 3.8.1 as well as dependencies using Vert.x.
It also updates the smallrye artifacts Ids that have been renamed.
Finally, it provides support for the back pressure control on emitter in Reactive Messaging.

It supersedes #3711.

cescoffier committed:
Update vert.x version to 3.8.1 as well as companions (Axle to 0.0.9, SR Reactive messaging to 1.0.3, SR reactive streams operators to 1.0.9)
Rename the artifactId of the SmallRye artifacts that went through the name refactoring.

cescoffier committed:
Extend emitter support to support back-pressure control

cescoffier committed:
Add missing Reactive Streams Ops dependency to the quarkus-vertx runtime

geoand commented:
JDK 11 tests seems to be failing

mkouba commented on extensions/smallrye-reactive-messaging/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/deployment/EmitterBuildItem.java:
```diff
@@ -9,6 +9,10 @@
      */
     private final String name;
 
+    private String overflow;
```
AFAIK build items should be immutable. What's the reason to make these fields mutable?

mkouba commented on extensions/smallrye-reactive-messaging/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/deployment/SmallRyeReactiveMessagingProcessor.java:
```diff
@@ -82,15 +84,24 @@ void validateBeanDeployment(
-                    emitters.produce(new EmitterBuildItem(name));
+                    EmitterBuildItem item = new EmitterBuildItem(name);
+                    if (annotation != null) {
+                        item.setOverflow(annotation.value().toString());
```
Ok now I see why those setters. Well, I'd personally prefer a static factory method for this particular use case, ie. something like `EmitterBuildItem.from(name, annotation)` and final fields.

cescoffier replied:
+1

cescoffier commented:
The CI fail was kind of expected as right now we have both Vert.x 3.80. (quarkus-http) and Vert.x 3.8.1. Working on it.

stuartwdouglas commented:
The version in the bom should override the version in quarkus-http, so unless there are some incompatibilities this should not be a factor.

cescoffier commented:
@stuartwdouglas in this release there is one incompatibility regarding the Cookie class moved from Vert.x web to Vert.x core.

gsmet commented:
@cescoffier in any case, as we import the Vert.x BOM, the versions should be consistent. If they aren't, it needs fixing in the BOM as we need them to be.

cescoffier commented:
@gsmet we need to exclude the Vert.x version from quarkus-http. Will do this.
Almost done with the quarkus-http release.

cescoffier committed:
Make the EmiterBuildItem immutable

cescoffier committed:
Update Quarkus-HTTP version to use Vert.x 3.8.1

cescoffier commented:
@mkouba I've made the `EmitterBuildItem` immutable.
@gsmet I've made sure the Vert.x version is the one from Quarkus
@stuartwdouglas I've released and bumped Quarkus HTTP

So... now the build should pass.... Fingers crossed. 


@tsegismont FYI

FroMage commented on bom/runtime/pom.xml:
```diff
@@ -904,6 +904,11 @@
                         <artifactId>netty-all</artifactId>
                     </exclusion>
+                    <exclusion>
+                        <!-- explicitly exclude the vert.x dependency to use the version define by Quarkus -->
```
s/define/defined/

FroMage commented on extensions/amazon-lambda/runtime/pom.xml:
```diff
@@ -45,6 +49,7 @@
             <artifactId>mockito-core</artifactId>
             <scope>test</scope>
         </dependency>
+
```
How's that related? Did this really pass the automated formatter?

cescoffier replied:
It actually did... weird.

cescoffier replied:
Removed the extra line.

FroMage commented on extensions/amazon-lambda/runtime/pom.xml:
```diff
@@ -35,6 +35,10 @@
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-vertx</artifactId>
+        </dependency>
```
Why the new dep?

cescoffier replied:
We exclude Vert.x from `quarkus-http`, so we need to have access to it.

FroMage replied:
OK.

gsmet replied:
Don't we need to add the deployment dependency in the deployment module?

FroMage commented on extensions/vertx/deployment/pom.xml:
```diff
@@ -34,7 +34,7 @@
             <artifactId>quarkus-vertx</artifactId>
         </dependency>
-	<dependency>
+	    <dependency>
```
Doubt this passes the automated formatter…

cescoffier replied:
Hum, I don't have this issue.
I had to fix it because it was using a tab instead of spaces.

FroMage replied:
Well, if you ran it, and it's fine…

gsmet replied:
The automated formatter doesn't check XML files, only Java.

But, yes, there is still a tab there.

FroMage commented on extensions/vertx/runtime/pom.xml:
```diff
@@ -47,6 +47,10 @@
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-smallrye-reactive-streams-operators</artifactId>
+        </dependency>
```
Why does the vert.x extension depend on SR-RSO now? I don't see any new code in that extension.

cescoffier replied:
It's to have access to the Axle classes. It's was merged in another PR. It should be fixed with the new Axle stuff.

FroMage replied:
I don't understand how SR-RSO is related to Axle.

cescoffier replied:
Axle provides methods returning `PublisherBuilder` defined in RSO.

FroMage replied:
Well, OK, sorry, but does the vert.x extension use `PublisherBuilder` now?

cescoffier replied:
It exposes the Axle Vert.x instance so somehow yes...
This dependency was added to be more convenient (in another PR, this is just a fix as the dependency for not declared in the runtime).

FroMage replied:
If it's exposed by the Axle Vert.x instance, it should be imported by the vertx axle module, no? If it's for convenience, that's another thing.

cescoffier replied:
Vert.x Axle depends on the spec only, not on the implementation. Here we explicitly depend on the implementation.

FroMage commented on test-framework/amazon-lambda/pom.xml:
```diff
@@ -31,5 +31,9 @@
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-vertx</artifactId>
+        </dependency>
```
Same question as for the lambda ext.

cescoffier replied:
Same answer, you need to give access to the Vert.x classes as they are not resolved by transitivity anymore.

mkouba commented on extensions/smallrye-reactive-messaging/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/deployment/SmallRyeReactiveMessagingProcessor.java:
```diff
@@ -82,15 +84,27 @@ void validateBeanDeployment(
                 if (stream != null) {
                     // Stream.value() is mandatory
                     String name = stream.value().asString();
+                    AnnotationInstance annotation = injectionPoint.getTarget().asField()
```
We should also support constructor and initializer/setter injection...

mkouba replied:
By the way, the `OnOverflow`'s javadoc is using wrong annotation name in the example: `OnBackPressure`.

cescoffier replied:
> We should also support constructor and initializer/setter injection...

what do you mean?

mkouba replied:
An injection point does not have to be a field ;-). You can also use a bean constructor or initializer methods (e.g. a setter) -> in that case, `InjectionPointInfo.getTarget()` returns a `MethodInfo`; `InjectionPointInfo.getPosition()` can be used to obtain the parameter annotations. `io.quarkus.arc.processor.Annotations.getParameterAnnotations(BeanDeployment, MethodInfo, int)` is a convenient util method to get the annotations including those added by means of `AnnotationsTransformer`.

cescoffier replied:
Oh I see...

cescoffier replied:
Ok should be fixed now.

cescoffier replied:
so it was NOT fixed.

cescoffier replied:
now it's fixed, and I've added a test to ensure it's working.

mkouba commented on extensions/smallrye-reactive-messaging/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/deployment/SmallRyeReactiveMessagingProcessor.java:
```diff
@@ -82,15 +84,27 @@ void validateBeanDeployment(
+                        emitters.produce(
+                                EmitterBuildItem.of(name,
+                                        annotation.value().toString(),
+                                        annotation.value("bufferSize").asInt()));
```
`annotation.value("bufferSize")` may return `null` if the value is not specified. You should either use `org.jboss.jandex.AnnotationInstance.valueWithDefault(IndexView, String)` or handle `null` value, e.g. with a hardcoded default value.

cescoffier replied:
Done.

gsmet commented on bom/runtime/pom.xml:
```diff
@@ -904,6 +904,11 @@
+                    <exclusion>
+                        <!-- explicitly exclude the vert.x dependency to use the version defined by Quarkus -->
+                        <groupId>io.vertx</groupId>
+                        <artifactId>vertx-core</artifactId>
```
What I was saying is that this shouldn't be necessary as we include the Vert.x bom at the beginning of this bom.

If it is an issue, then there's something going wrong.

cescoffier replied:
Your BOM won't change the version imported by a dependency.

cescoffier replied:
I may have a solution, but that may end up ugly.

cescoffier replied:
Actually nope.

What could be done is to have an _empty_ module in quarkus that depends on Vert.x core and quarkus-http (in this order), and all other modules would depend on this module.

As there is no easy solution, we probably should remove the `exclusion` and manage the alignment by hand for now. That would remove the dependencies added to the lambda modules.

gsmet replied:
> Your BOM won't change the version imported by a dependency.

Hu? Are we talking about the same thing? A BOM is part of dependency management so it will apply a consistent version across all your dependencies.

cescoffier replied:
Here, quarkus-http was depending on vert.x 3.8.0.
Even if the bom says that the Vert.x version is 3.8.1, if a module depends on quarkus-http, and does not explicitly depends on vert.x, it will resolve to 3.8.0.

cescoffier replied:
That was actually the first build issue. Even with the BOM, it was resolving to Vert.x 3.8.0.

cescoffier replied:
I've removed the exclusion.

gsmet replied:
Then we have an issue with the BOM because it should enforce consistently the Vert.x version across all the codebase, including in your case. That's what dependency management is for. So if it doesn't work, we have something to fix with the Vert.x BOM inclusion.

cescoffier replied:
Then we need to merge quarkus-http in the main codebase or implement the _empty_ module trick.

gsmet replied:
No, it should work as is. That's what I'm trying to say. Dependency management will enforce consistency across all your dependencies and the dependencies of your dependencies (and so on).

I trust you that there was an issue, I'll check later what's going on with the BOM.

cescoffier replied:
So right now, it enforce the vert.x version in the quarkus codebase which is what we want.

cescoffier committed:
Handle other kind of injection point (method, parameter...)
Handle hwne bufferSize is not set

cescoffier committed:
Add support for emitter injection in methods and constructors

cescoffier committed:
Bump netty version to 4.1.39.Final

cescoffier commented:
It will likely not work, as the updated netty contains substitutions and so we have during duplicates. I will investigate tomorrow.

cescoffier commented:
ok, the missing dependency has finally been cached by the Google Maven mirror.

cescoffier committed:
Remove substitution contained in the upstream netty artifact

cescoffier committed:
Fix Netty substitutions related to SSL

cescoffier committed:
Delay the initialisation of the VertxHttp2ClientUpgradeCodec to avoid a GraallVM compilation issue.

This issue was rasied when the application uses a WebClient.

cescoffier committed:
Temporary disable the async client of dynamoDB until the AWS SDK becomes compatible with the netty version we ship.
More details on https://github.com/aws/aws-sdk-java-v2/issues/1399, https://github.com/aws/aws-sdk-java-v2/issues/1344 and https://github.com/aws/aws-sdk-java-v2/pull/1391

cescoffier commented:
Update:

* Another issue with the Vert.x web client has been fixed
* Due to AWS SDK incompatibility with the new netty, I've disabled the async support of the dynamodb extension. An issue has been filled on the SDK (https://github.com/aws/aws-sdk-java-v2/issues/1399), and the @marcinczeczko has been made aware (and is ok with it)
